network/src/protocols/health_checker/mod.rs

```diff
@@ -194,72 +194,72 @@ where
                             };
                         }
                         Event::Message(peer_id, msg) => {
                             error!(
                                 SecurityEvent::InvalidNetworkEventHC,
-                                NetworkSchema::new(&self.network_context)
-                                    .remote_peer(&peer_id),
+                                NetworkSchema::new(&self.network_context).remote_peer(&peer_id),
                                 "{} Unexpected direct send from {} msg {:?}",
                                 self.network_context,
                                 peer_id,
                                 msg,
                             );
                             debug_assert!(false, "Unexpected network event");
-                        },
+                        }
                     }
                 }
                 _ = self.ticker.select_next_some() => {
                     self.round += 1;
-                    match self.sample_random_peer() {
-                        Some(peer_id) => {
-                            let nonce = self.sample_nonce();
-                            trace!(
-                                NetworkSchema::new(&self.network_context),
-                                round = self.round,
-                                "{} Will ping: {} for round: {} nonce: {}",
-                                self.network_context, peer_id.short_str(),
-                                self.round,
-                                nonce
-                            );
+                    let peers: Vec<_> = self.connected.keys().cloned().collect();
 
-                            tick_handlers.push(
-                                Self::ping_peer(
-                                    self.network_context.clone(),
-                                    self.network_tx.clone(),
-                                    peer_id,
-                                    self.round,
-                                    nonce,
-                                    self.ping_timeout));
-                        }
-                        None => {
-                            trace!(
-                                NetworkSchema::new(&self.network_context),
-                                round = self.round,
-                                "{} No connected peer to ping round: {}",
-                                self.network_context,
-                                self.round
-                            );
-                        }
+                    if peers.is_empty() {
+                        trace!(
+                            NetworkSchema::new(&self.network_context),
+                            round = self.round,
+                            "{} No connected peer to ping round: {}",
+                            self.network_context,
+                            self.round
+                        );
+                    }
+
+                    for peer_id in peers {
```

Comment:
`for peer_id in self.connected.keys() {`

Reply:
Without some rust magic (that i'm not quite aware of yet), I can't do that.

`self.connected.keys()` is an immutable borrow, then `self.nonce()` is a mutable borrow.  So, I'm leaving it with the cloning for now.